Retain the cursor position when transforming the text

hartez opened the pull request:
### Description of Change

When an Entry has a text transform applied, the managed cursor position tracking algorithm fails and the cursor is reset to 0. This change uses the native Android method for setting the text and preserving the cursor state.

Update:

Text input updates in Controls have two different situations they need to handle. When the cross-platform property is updated, it needs to set the text in the control and move the cursor position to the end. But when the platform control is updating the cross-platform property, we _don't_ want to force the cursor to the end; we want it to stay in place. And in some situations, we also need to update the text in place (e.g., when using a TextTransform) while preserving the cursor location. 

So the mappings need a way to determine which is happening. These changes add a mechanism to the handlers which can indicate when the data is flowing cross-platform -> platform (the default), and when it's flowing the other way (from the platform control to the cross-platform property). 

### Issues Fixed

Fixes #11313

hartez committed:
Retain the cursor position when transforming the text
Fixes #11313

rmarinho reviewed (changes requested):
Seems tests are failing on Android

hartez committed:
Allow the handler to signal the direction of information so the mappings
can handle things differently when setting properties from xplat and updating
xplat properties from controls;

hartez committed:
Fix TextTransform cursors for Editor

hartez committed:
Fix TextTransform cursors for Editor

hartez committed:
Add tests

mattleibow commented:
I both like and worry about this property. Two things come to mind. 

If text changes in the platform, then this is set... And then setting something triggers other properties. Will it cause this to be set and unset?

Another thing is... I took a break typing this and forgot... Something to do with 1 property controlling all the properties... Will update when it comes back to me...

hartez commented:
> If text changes in the platform, then this is set... And then setting something triggers other properties. Will it cause this to be set and unset?

Not sure I understand what you're asking, but yes, there is a potential situation where 
1. The user types something into an Entry
2. The flag is set
3. The xplat text is updated, and the user has added a handler for xplat TextChanged
4. During the time that the TextChanged handler is doing its work, the flag is still set
5. Everything finishes, and the flag is cleared

Stuff happening during 4 may cause confusion/problems if the user is using the Entry's TextChanged event to modify stuff about the Entry.

> Something to do with 1 property controlling all the properties... 

As an alternative, we could add property-specific flags to the handlers for every property where we have this issue. But generally speaking, controls don't tend to have multiple "input" properties; I think we'd just end up with one flag per input control anyway.

If anyone's got a better option than setting flags to accomplish this, I'd love to see it. I don't love the flag for this, it's just the only way I've found to make it work.
